### PR TITLE
Unsloth optims for Llama

### PR DESCRIPTION
**Pull request by winglian**
WIP to integrate Unsloth's optimizations into axolotl. 

The manual autograd for MLP, QKV, O only seems to help VRAM by 1% as opposed to the reported 8%. 
The Cross Entropy Loss does help significantly, but only reduced VRAM by 13% as opposed to the reported 17%.

edit: clarification, the cross entropy loss works for both full fine tunes and lora. The MLP, QKV, and O are only for 4-bit qlora with flash attention.

### CODE REVIEW

**Comment by bratao** [comment]
The cross_entropy_loss optimization is applicable even in a full fine tune, right?

**Comment by winglian** [comment]
> The cross_entropy_loss optimization is applicable even in a full fine tune, right?

Correct!

**Comment by linux-leo** [comment]
Are these optimizations compatible with flash attention? (Complete Noob here)

**Comment by bratao** [comment]
It is possible to alter to also patch Qwen? It was added to unsloth and all optimizations work for it:
https://github.com/unslothai/unsloth/pull/428

**Comment by winglian** [comment]
> It is possible to alter to also patch Qwen? 

Let's tackle that in a follow up PR

**Comment by Nero10578** [comment]
Does this only work for single GPU? Getting error of "Runtime Error: Model must be 2-D" when enabling unsloth cross entropy loss on 2x3090Ti finetuning Llama 3 8B LORA.
